pkg/ovs/ovsctl/appctl_test.go

```diff
@@ -99,14 +131,152 @@ IPv6 ND Extension: No`,
 			got, err := c.GetDPFeatures()
 			require.NoError(t, err)
 			assert.Equal(t, tt.want, got)
 		})
 	}
 }
+
+func TestTrace(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockOVSAppctlRunner := NewMockOVSAppctlRunner(ctrl)
+	srcMAC, _ := net.ParseMAC("11:11:11:11:11:11")
+	dstMAC, _ := net.ParseMAC("11:11:11:11:11:22")
+	srcIP := net.ParseIP("1.1.1.2")
+	dstIP := net.ParseIP("2.2.2.2")
+	tc := []struct {
+		name          string
+		req           *TracingRequest
+		expectedFlow  string
+		expectedError string
+	}{
+		{
+			name: "Trace IP Packet-success",
+			req: &TracingRequest{
+				InPort:              "3",
+				SrcIP:               srcIP,
+				DstIP:               dstIP,
+				SrcMAC:              srcMAC,
+				DstMAC:              dstMAC,
+				Flow:                "in_port=3,tcp,tcp_dst=22,",
+				AllowOverrideInPort: true,
+			},
+			expectedFlow: "table=0,in_port=3,ip,nw_src=1.1.1.2,action=resubmit(,2)\ntable=2,ip,tcp,tcp_dst=22,action=output:1\n",
+		},
+		{
+			name: "no-ops due to source and destination IP error for non-IP packet",
+			req: &TracingRequest{
+				InPort: "9",
+				SrcIP:  srcIP,
+				Flow:   "table=10, n_packets=0, n_bytes=0, priority=200,arp,actions=resubmit(,20)",
+			},
+			expectedError: "source and destination must not be specified for non-IP packet",
+		},
+		{
+			name: "no-ops due to duplicated source IP error",
+			req: &TracingRequest{
+				InPort: "9",
+				SrcIP:  srcIP,
+				Flow:   "nw_src=1.1.1.2,tcp,tcp_dst=22,",
+			},
+			expectedError: "duplicated 'nw_src' in flow",
+		},
+		{
+			name: "no-ops due to duplicated destination IP error",
+			req: &TracingRequest{
+				InPort: "9",
+				DstIP:  dstIP,
+				Flow:   "nw_dst=2.2.2.2,tcp,tcp_dst=22,",
+			},
+			expectedError: "duplicated 'nw_dst' in flow",
+		},
+		{
+			name: "no-ops due to duplicated in_port error",
+			req: &TracingRequest{
+				InPort:              "9",
+				Flow:                "in_port=3,tcp,tcp_dst=22,",
+				AllowOverrideInPort: false,
+			},
+			expectedError: "duplicated 'in_port' in flow",
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			runTracing := func() ([]byte, *ExecError) {
+				var x string
+				for i := 0; i < len(testFlows); i++ {
+					if strings.Contains(testFlows[i], fmt.Sprintf("nw_src=%s", tt.req.SrcIP.String())) {
+						x = testFlows[i] + "\r\n"
+						break
+					}
+				}
+				tcpdst := strings.Split(tt.req.Flow, "tcp_dst=")[1]
+				for i := 0; i < len(testFlows); i++ {
+					if strings.Contains(testFlows[i], fmt.Sprintf("tcp_dst=%s", tcpdst)) {
+						x = x + testFlows[i] + "\r\n"
+						break
+					}
+				}
+				result := []byte(x)
+				return result, nil
+			}
+			client := &ovsCtlClient{
+				bridge:          "br-int",
+				ovsAppctlRunner: mockOVSAppctlRunner,
+			}
+			if tt.name == "Trace IP Packet-success" {
+				mockOVSAppctlRunner.EXPECT().RunAppctlCmd("ofproto/trace", true, gomock.Any()).Return(runTracing())
```

Comment:
you have this a bit backwards: you should validate that the correct parameter is passed to `RunAppctlCmd` instead of using `gomock.Any()`; on the other hand, the return value is not very important given that the `Trace` function doesn't perform any significant transformation on the result of `ovs-appctl ofproto/trace` (hence you can remove `runTracing` altogether).